Add stress test for 3d and combine in one function

henrij22 opened the pull request:
This is a small PR adding a stress test for linear 3D cube elements .
We should discuss for a different PR how and if we want to modify this whole testing framework, maybe storing the expected values in the element formulation or sth like that.

For reference: my calulations come from these notebooks: https://github.tik.uni-stuttgart.de/HenrikJakob/elements-notebooks/tree/main

I might also add a test for 3d cube with EAS tommorow, so we can also add this to this PR

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -187,139 +186,88 @@ template <typename NonLinearOperator>
+                "element dimension");
+  assert(fe.localView().element().type() == Dune::GeometryTypes::cube(dim)
+         or fe.localView().element().type() == Dune::GeometryTypes::simplex(dim)
+                && "The test to check linear stress is only supported for cubes");
```
The assert message is out of sync.

henrij22 replied:
Ah yeah sorry, I was mindlessly copy-pasting these

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -187,139 +186,88 @@ template <typename NonLinearOperator>
       "The test to check linear stress is only supported for a linear Lagrange basis");
 
-  Dune::TestSuite t("Linear Stress check of" + std::to_string(dim) + "D  simplex element");
+  Dune::TestSuite t("Linear Stress check of" + std::to_string(dim) + "D  cube element");
```
This is also out of sync?

rath3t replied:
You can get the name via https://gitlab.dune-project.org/core/dune-geometry/-/blame/master/dune/geometry/type.hh?ref_type=heads#L421

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -187,139 +186,88 @@ template <typename NonLinearOperator>
+  using namespace Ikarus;
+  auto gridView                = fe.localView().globalBasis().gridView();
+  auto element                 = elements(gridView).begin();
+  const auto& referenceElement = Dune::ReferenceElements<double, dim>::general(element->type());
```
you can also get the reference Element via   `referenceElement(basis.localView().tree().element())` you can pass this further above

henrij22 replied:
Ah i even think this is not neccesary, we could something like that (out of my head):
```cpp
auto element = fe.getGridElement();
const auto& refElement =Dune::referenceElement(element);
```

rath3t replied:
Yes thats right. Basically you just need a grid element from somewhere.

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -187,139 +186,88 @@ template <typename NonLinearOperator>
-template <typename NonLinearOperator, typename FiniteElement>
-[[nodiscard]] auto checkLinearStressOfSimplex(NonLinearOperator& nonLinearOperator, FiniteElement& fe,
-                                              const std::string& messageIfFailed = "") {
+[[nodiscard]] auto checkLinearStress(NonLinearOperator& nonLinearOperator, FiniteElement& fe,
```
Nice generalization.
I still would aim for a further abstraction level here.
I envision the signature of this function more like:

```cpp
template <typename NonLinearOperator, typename FiniteElement>
[[nodiscard]] auto checkCalculateAt(NonLinearOperator& nonLinearOperator, FiniteElement& fe,const ResultRequirements& resultRequirements,  std::vector<Eigen::Matrix...>& expectedResults,  const std::vector<LocalCoordinates>& localPositionsOfResults
                                              const std::string& messageIfFailed = "") 
```

Now we can pass in from the outside which results we require and we can also set global solutions:

```cpp

  displacement << 0, 0, 1, 1, 1, 1, 1, 1;
  std::array<Dune::FieldVector<double, 3>, 4> expectedStress = {{{1428.5714287000, 1428.5714287000, 769.2307693000},
                                                                 {1098.9010990000, 329.6703297000, 384.6153846500},
                                                                 {329.6703297000, 1098.9010990000, 384.6153846500},
                                                                 {0, 0, 0}}};
  auto resultRequirements = ResultRequirements()
                                .insertGlobalSolution(FESolutions::displacement, displacement)
                                .addResultRequest(ResultType::linearStress);

std::array<LocalCoordinate,4> positions;
for (auto i : std::views::iota(0, nodes)) 
    positions[i] = referenceElement.position(i, dim);

checkCalculateAt(nonLinOp,fe,resultRequirements,expectedStress ,positions);
```

To really support std::array and std::vector for expected results and coordinate position in one function, you can look int [std::span<T,N>](https://en.cppreference.com/w/cpp/container/span), which allows static and dynamic extend

rath3t replied:
In this way, i would hope that we never have to touch this function again if we want to check further results of other elements. This nicely is in line with the [Open–closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle)

rath3t replied:
Excepte we refactor the calculateAt signature obviously

rath3t replied:
Furthermore, at the outside this leads to some code duplication but i think you can write then also a wrapper.
like createLinearStressResults or something, which return you `resultRequirements,expectedStress ,positions`.

by aware you can do this via structured bindings.

```cpp
auto createLinearStressResults(...)
{
//...
return std::make_tuple(resultRequirements,expectedStress ,positions);
}
```

and at the call site this would look like

```cpp
const auto [resultRequirements,expectedStress ,positions] =createLinearStressResults(...); //directly un-pack tuple
checkCalculateAt(nonLinOp,fe,resultRequirements,expectedStress ,positions);
```

rath3t replied:
At https://github.com/ikarus-project/ikarus/blob/7e793d0ae955b4b11ba7d2c98522775495abe703/tests/src/testlinearelasticity.cpp#L43C52-L43C52 you now have to provide your own functor.

Which can be obtained via:

```cpp

auto checkCalculateAtFactory(const auto& resultRequirements,const auto& expectedStress ,const auto& positions)
{
return [=](auto& nonLinOp, auto& fe, [[maybe_unused]] auto& req) { return checkCalculateAt(nonLinOp, fe,resultRequirements,expectedStress ,positions); };
}
```

There i think we have to capture the arguments by value instead of reference, since later these could by dangling. This is done by the lambda capture `[=]`

rath3t reviewed (changes requested):
I hope this is not going to crazy here. See comments and let me know if something would break.

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -187,139 +186,88 @@ template <typename NonLinearOperator>
+  for (auto i : std::views::iota(0, nodes)) {
+    auto pos = referenceElement.position(i, dim);
+    fe.calculateAt(resultRequirements, pos, result);
+    computedStress(i, Eigen::placeholders::all) = result.getResult(ResultType::linearStress).transpose();
```
Ah and I forgot. In this case here i would write 
```cpp
computedStress.row(i) = result.getResult(ResultType::linearStress).transpose();
``` 
since i think this makes your intent more clear. Maybe even the transpose is not needed then.

henrij22 replied:
Ah yeah i wanted to change it to `row(i)` but i forgot, the `transpose()` is actually needed none the less

rath3t commented on tests/src/resultcollection.hh:
```diff
@@ -0,0 +1,72 @@
+    if constexpr (dim == 2) {
+      displacement << 0, 0, 2, 0, 1, 0;
+      constantStresses << 2197.80219780, 659.34065934, 384.61538462;
+      expectedStress << constantStresses, constantStresses, constantStresses;
```
Just to let you know about the functionality. You could have written here:
```cpp
expectedStress.rowwise() = constantStresses;
```
see https://eigen.tuxfamily.org/dox-3.2/group__TutorialReductionsVisitorsBroadcasting.html section Broadcasting

but you can decide what is more readable.

henrij22 replied:
ah i was looking for this kind of functionality but couldn't find it

rath3t commented on tests/src/resultcollection.hh:
```diff
@@ -0,0 +1,72 @@
+  const int nodes              = referenceElement.size(dim);
+
+  Eigen::MatrixXd expectedStress(nodes, dim * (dim + 1) / 2);
+
```
You can also move `expectedStress` inside the if statements and then return inside the constexpr. 
Therein you have all the information for the sizes of the result anyway and then you can get rid of the need for `nodes`. In my point of view, there should be one functor for each test.

So more like

```cpp
auto createLinearTriangleStressExpectedResults()
{
  Eigen::VectorXd displacement;
  displacement << 0, 0, 2, 0, 1, 0;

  auto resultRequirements = ResultRequirements()
                                .insertGlobalSolution(FESolutions::displacement, displacement)
                                .addResultRequest(ResultType::linearStress);
using ReferenceElements = Dune::ReferenceElements<double, 2>;
const auto& referenceElementTriangle = ReferenceElements::simplex();

  std::vector<typename ReferenceElements::Coordinate > positions;
  for (auto i : std::views::iota(0, nodes))
    positions.push_back(referenceElementTriangle.position(i, 2));

Eigen::MatrixXd expectedStress(3, 3);
 expectedStress.rowwise() << 2197.80219780, 659.34065934, 384.61538462; // dont know if this really works
 

return std::make_tuple(resultRequirements, expectedStress, positions);
}

```

This add code duplication but i thinks its much more readable and less if else branches.

Furthermore, you can write a helper function to get all nodal positions, which can be handy anyway.
This would reduce the code to


```cpp
auto createLinearTriangleStressExpectedResults()
{
  Eigen::VectorXd displacement;
  displacement << 0, 0, 2, 0, 1, 0;

  auto resultRequirements = ResultRequirements()
                                .insertGlobalSolution(FESolutions::displacement, displacement)
                                .addResultRequest(ResultType::linearStress);

Eigen::MatrixXd expectedStress(3, 3);
 expectedStress.rowwise() << 2197.80219780, 659.34065934, 384.61538462; // dont know if this really works

return std::make_tuple(resultRequirements, expectedStress, getVertexPositions(Dune::GeometryTypes::triangle));
}

```

henrij22 replied:
Okay so you want for each Elementshape + dimension a different function that fives the expectedResult? What about EAS?

- [ ] Q1 2D
- [ ] Q1 3D
- [ ] Q1 2D EAS
- [ ] T1 2D
- [ ] T1 3D

rath3t replied:
Yes so basically seperating `createLinearStressResults` for each test case as stated above.
For EAS i think we still have to pass the fe, since we have to call setEASType. which is unfortunate :/. But I don't see way around it currently.

henrij22 replied:
Okay, i would prefer the functions to all have the same interface, so all taking the fe argument, what do you think?
E.g.

```cpp
auto createLinearStressResultsForSimplex = []<typename FiniteElement>(FiniteElement& fe) {}
```
or as we don't really need the type this would be enough
```cpp
auto createLinearStressResultsForSimplex = [](auto& fe) {}
```

rath3t replied:
yes i think all should accept the fe argument

henrij22 replied:
Also i need the nonlinop as of now, because when I do that
```cpp
Eigen::VectorXd displacement;
displacement << 0, 0, 2, 0, 1, 0;
.insertGlobalSolution(FESolutions::displacement, displacement)
```
the `displacment` goes out of scope and is not longer available in the test function. Thats why earlier iteration of this were doing 
```cpp
  auto& displacement = nonLinearOperator.firstParameter();
```
The next iteration will be ready in a few minutes

rath3t replied:
You can overload https://github.com/ikarus-project/ikarus/blob/7e793d0ae955b4b11ba7d2c98522775495abe703/ikarus/finiteelements/mechanics/linearelastic.hh#L32

and the same way here: https://github.com/ikarus-project/ikarus/blob/7e793d0ae955b4b11ba7d2c98522775495abe703/ikarus/finiteelements/ferequirements.hh#L117

such that `SolutionVectorType_ ` is `Eigen::VectorXd` instead of a reference to it.

But since also passing the nonLinearOperator anyway I'm fine with the current solution, but anyway it would be nice anyway. But I'm not sure if this would be worth it.

henrij22 replied:
Yeah i think its not too bad to have the nonLinOp as a function argument as its already available in these test functors

rath3t reviewed (changes requested):
Sorry to bother you further with another iteration. See comments

henrij22 committed:
Add stress test for 3d and combine in one function

+ housekeeping
add asserts

fix clang format

WTF did i do now?

(fixed)

merge all stress tests into one function

Fixed digits of t1 test

Eas values

further generalization

2nd iteration

rath3t commented on tests/src/resultcollection.hh:
```diff
@@ -0,0 +1,170 @@
+  constexpr int dim            = FiniteElement::Traits::mydim;
+  const auto& element          = fe.gridElement();
+  const auto& referenceElement = Dune::referenceElement<double, dim>(element.type());
+  const int nodes              = referenceElement.size(dim);
```
can you rename this to `numberOfVertices` since this is information of the reference element and not of a basis.

In the dune context nodes are quantities of the basis. Therefore, a square with quadratic ansatz has 4 vertices but 9 nodes.

henrij22 committed:
wording and get rid of comment

rath3t commented on tests/src/resultcollection.hh:
```diff
@@ -0,0 +1,170 @@
+  const int nodes              = referenceElement.size(dim);
+
+  std::vector<typename FiniteElement::GridElementEntityType::Geometry::LocalCoordinate> positions;
+  for (auto i : std::views::iota(0, nodes))
```
You can also write here:

```cpp
std::ranges::transform(std::views::iota(0, nodes), positions, std::back_inserter(positions), [&](int i){return referenceElement.position(i, dim);})
```

Since I have the impression you like ranges and i.e. `std::views::iota` ;)

henrij22 replied:
Haha i think in this case, the non range variant is a bit more clear :D

rath3t commented on tests/src/resultcollection.hh:
```diff
@@ -0,0 +1,170 @@
+    expectedStress << 1214.28571429, 1214.28571429, 384.61538462, 1214.28571429, 214.28571429, 384.61538462,
+        214.28571429, 1214.28571429, 384.61538462, 214.28571429, 214.28571429, 384.61538462;
+  } else {
+    expectedStress << 1428.57142857, 1428.57142857, 769.23076923, 1098.90109890, 329.67032967, 384.61538462,
```
Strange that there are these zeros at the end for non eas. For `setEASType(4)` the stress is constant within the element but for no EAS its variable. @tarun-mitruka can you comment. (This is not blocking the PR)

henrij22 replied:
Yeah thats interesting, the last 3 digits are representing the stresses on the upper right vertex (dune ordering of vertices)
I also recreated these results with my own notebooks.

tarun-mitruka replied:
I think since it is a single element test, the top right corner sees it as a rigid body translation (see deformed configuration in blue below) and shows zero stresses. I am assuming it should give better results with mesh refinement and thereby the solution from EAS is more correct. Does this make sense? This has to be investigated further though. The solution for a single element without EAS for shear strain for example is $2-\xi-\eta$ whereas EAS4 adds additional strains of $-1+\xi+\eta$ making the shear strain as $1$ in the whole element.
Also, the stress values written here only have 10 digits after the decimal point. Do you think we should add more digits and check it against a tolerance of 1e-15?

![image](https://github.com/ikarus-project/ikarus/assets/92575489/f8991805-bfed-4dac-a0b7-21617e701da3)

rath3t commented on tests/src/resultcollection.hh:
```diff
@@ -0,0 +1,170 @@
+};
+
+//
+// auto createLinearStressResults = []<typename NonLinearOperator, typename FiniteElement>(
```
I would remove all the commented code here before merging

henrij22 replied:
Already did that

rath3t replied:
Ah ok. Thanks.

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -187,139 +186,28 @@ template <typename NonLinearOperator>
+[[nodiscard]] auto checkCalculateAt(NonLinearOperator& nonLinearOperator, FiniteElement& fe,
+                                    const Ikarus::ResultRequirements<>& resultRequirements, const auto& expectedResult,
+                                    const auto& evaluationPositions, const std::string& messageIfFailed = "") {
+  Eigen::MatrixXd computedResults(expectedResult.rows(), expectedResult.cols());
```
You can also write here:

```cpp
Eigen::MatrixXd computedResults;
computedResults.setZero(expectedResult.rows(), expectedResult.cols());
```

or 

```cpp
Eigen::MatrixXd computedResults= Eigen::MatrixXd::Zero(expectedResult.rows(), expectedResult.cols());
```

or not setting it zero any since it will be overwritten anyway. Setting the size will be enough.
```cpp
Eigen::MatrixXd computedResults(expectedResult.rows(), expectedResult.cols());
```

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -187,139 +186,28 @@ template <typename NonLinearOperator>
-                           << "It is\n:" << computedResult << "\nBut should be:\n"
-                           << expectedStress << "\n"
+  Dune::TestSuite t("Test of the calulateAt function");
+  const bool isStressCorrect = isApproxSame(computedResults, expectedResult, 1e-8);
```
can you rename to isResultCorrect.

Furthermore, in the error message you can write somewhere `toString(resultRequirements.getRequestedResult()` to have a better error message, because then you know the actual result type that failed.

rath3t commented on tests/src/testfeelement.hh:
```diff
@@ -135,10 +127,14 @@ inline auto checkJacobianFunctor = [](auto& nonLinOp, [[maybe_unused]] auto& fe,
     = [](auto& nonLinOp, auto& fe, auto& req) { return checkFEByAutoDiff(nonLinOp, fe, req); };
+
+auto checkCalculateAtFunctorFactory(const auto resultCollectionFunction) {
+  return [=](auto& nonLinOp, auto& fe, [[maybe_unused]] auto& req) {
```
The way this is constructed here, I think it is sufficient to capture by reference `&` isn't it?

henrij22 replied:
It actually is.

rath3t reviewed (changes requested):
Final refactor. I swear! :D

henrij22 committed:
Req Review changes part 999 ;)
